### PR TITLE
Allow partial deserialization of molecules

### PR DESCRIPTION
**Pull request by greglandrum**
Sometimes you want to deserialize a molecule to work with it but you don't need all the stored information like conformers or properties. The idea here is allow you to skip reading in conformers and/or properties.

This PR adds an optional argument to the depickler calls which allows conformers and/or properties to be ignored when deserializing or when constructing a molecule from a pickle.

I also modified the pickling code to allow conformers to be skipped when serializing.

To make this work we need to add some additional info to the pickles, so this includes a major version bump of the pickle format. Backwards compatibility with old pickles is, of course, maintained.

I also explored adding the option to not deserialize the ring information. This didn't have an impact on deserialization performance on either the Mac or Linux system I tested it on, so that's not in this PR

Here's some benchmarking data for 100K Pubchem molecules read from an SDF and serialized with `PicklerProps::AllProps`:

- Parsing the molecules from the SDF: 20.8s
- Serializing the molecules (including properties and conformer data): 5.0s
- Normal deserialization: 3.1s
- Deserializing without properties: 2.1s
- Deserializing without properties or conformers: 1.0s
- Deserializing without conformers: 3.4s

I have not been able to figure out why it takes longer to ignore the conformer data than it does to parse it, but this is pretty consistent on my machine.

Further notes on performance. I duplicated that experiment on my Mac (M1 Mac Mini):
- Parsing the molecules from the SDF: 20.2s
- Serializing the molecules (including properties and conformer data): 5.0s
- Normal deserialization: 3.1s
- Deserializing without properties: 1.3s
- Deserializing without properties or conformers: 0.9s
- Deserializing without conformers: 2.5s

So the odd problem with skipping conformers being slower than reading them is not observed with this compiler/hardware combo.

### CODE REVIEW

**Review by d-b-w — COMMENTED** [review]
I wonder if it would make sense to store the stuff that we don't expand as a string property.

That would allow workflows like - scan my so many molecules for some interesting feature. Once I found the ones I like, get their conformers and other data.

**Comment by greglandrum** [comment]
> I wonder if it would make sense to store the stuff that we don't expand as a string property.
> 
> That would allow workflows like - scan my so many molecules for some interesting feature. Once I found the ones I like, get their conformers and other data.

We'd need to also expose functionality allowing that data to be deserialized in place. You can achieve a similar effect by doing the partial deserialization before doing the scan and then just re-parsing the pickle if the molecule actually is interesting?

File: Code/GraphMol/MolPickler.cpp [hunk 1]
```diff
@@ -1046,53 +1058,67 @@ void MolPickler::_pickle(const ROMol *mol, std::ostream &ss,
...
-      const Conformer *conf = ci->get();
-      _pickleConformer<T, float>(ss, conf);
+    if (propertyFlags & PicklerOps::MolProps) {
+      streamWrite(tss, BEGINCONFPROPS);
+      std::stringstream tss2;
+      for (auto ci = mol->beginConformers(); ci != mol->endConformers(); ++ci) {
+        const Conformer *conf = ci->get();
+        _pickleProperties(tss2, *conf, propertyFlags);
+      }
+      write_sstream_to_stream(tss, tss2);
```
[the remark this answers is not in the document]
Comment (bp-kelley) [reply Code/GraphMol/MolPickler.cpp:1090]: These strings could be stored on the molecule, as long as the bits are tagged, we might be able to expose a simple api too get the rest of the data.

[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/MolPickler.cpp:1090]: Again, I don't think this is an actual performance problem, but we can add something for it later if it turns out to matter

**Comment by d-b-w** [comment]
I think that you're right and that we don't need to store the unparsed bits. I was thinking of some circumstance in which the caller would read through oodles of these and then find the couple of interest, but I think that if someone is using a serialized mol, they can probably find the serialized data fairly efficiently.

File: Code/GraphMol/MolPickler.cpp [hunk 2]
```diff
@@ -94,6 +94,15 @@ std::mutex &GetPropMutex() {
   return propmutex_get();
 }
 #endif
+
+void write_sstream_to_stream(std::ostream &outStream,
+                             const std::stringstream &toWrite) {
+  auto ts = toWrite.str();
```
Comment (bp-kelley) [inline Code/GraphMol/MolPickler.cpp:100]: Is there a better way to get the size of a string stream?  Perhaps seek(0) then tellg?

File: Code/GraphMol/MolPickler.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (bp-kelley) [reply Code/GraphMol/MolPickler.cpp:1090]: Following up with @d-b-w, the new format is basically:

```
<tag><size><payload>
```

which allows skipping.  It also allows reordering  and skipping of unknown tags which can be useful for backwards compatibility.  (it also allows xml style tags in tags).  Am I reading this correctly?  If so, we may consider a future reader where the tag orders don't matter as much, i.e. a switch statement.

[the remark this answers is not in the document]
Comment (bp-kelley) [reply Code/GraphMol/MolPickler.cpp:1090]: Actually it also allows attaching writers and readers unknown to the pickler.  I.e. if you wanted to stuff a grid onto a molecule or say visualization props.

**Review by bp-kelley — APPROVED** [review]
If I'm right about the the format, I think this is a good start for a more flexible pickler.

File: Code/GraphMol/MolPickler.cpp (hunk 2 above)
Comment (greglandrum) [reply Code/GraphMol/MolPickler.cpp:100]: we need to get the string anyway in order to write it out

File: Code/GraphMol/MolPickler.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/MolPickler.cpp:1090]: > which allows skipping. It also allows reordering and skipping of unknown tags which can be useful for backwards compatibility. (it also allows xml style tags in tags). Am I reading this correctly? If so, we may consider a future reader where the tag orders don't matter as much, i.e. a switch statement.

Yeah, we could definitely use this to introduce a more extensible-while-maintaining-backwards-compatibility pickle format. There are some things that need to be done in a particular order (like atoms and bonds first), but that's not a big deal.

File: Code/GraphMol/MolPickler.cpp (hunk 2 above)
Comment (bp-kelley) [reply Code/GraphMol/MolPickler.cpp:100]: I think you can do

```
outStream << toWrite;
```

But if it's a premature optimization, that's ok.

**Review by bp-kelley — APPROVED** [review]

**Pull request merged**